inst/include/conditionalSampler.h

```diff
@@ -0,0 +1,673 @@
+                    dRSWeightsResidual.at(i) = dRSWeightsResidual.at(i)/arma::sum(dRSWeightsResidual);
+                    //Deterministic component assignment
+                    for (int j : DsetCurrent) {
+                        uRSIndices.at(j) = i;
+                    }
+                    //Convenient way of handling the conditioning path
+                    //if(i == K_{t-1}) store D_{K_{t-1}};
+                    if(i == static_cast<int>(referenceTrajectoryIndices.at(T))) {
+                        DsetKtMinus1 = DsetCurrent;
+                    }
+                }
+                //Step 2:
+                //Sample K_t from appropriate lambda distribution
+                //    2.1 Initializes container
+                arma::Col<double> lambdaProbs(N);
+                double prob1 = 1/(N * dRSWeights.at(referenceTrajectoryIndices.at(T)));
+                double prob2 = dRSWeightsResidual.at(referenceTrajectoryIndices.at(T))/(N * dRSWeights.at(referenceTrajectoryIndices.at(T)) * (N - l));
+                //Defines the complement set of DsetKtMinus1 relative to {1,...,N} i.e. {1,...,N}\DsetKtMinus1
+                //    2.2 Fills a vector with probabilities to sample from \lambda(k_t|k_{t - 1}, x_{t - 1}^{1:N})
+                for (int n = 0; n < N; ++n) {
+                    if (any(arma::find(DsetKtMinus1 == n))) {
+                        lambdaProbs.at(n) = prob1;
+                    } else {
+                        lambdaProbs.at(n) = prob2;
+                    }
+                }
+                //    2.3 Sample Kt from eq. 4 on p.8 of the WP
+                long Kt = Rcpp::sample(N, 1, false, Rcpp::wrap(lambdaProbs))[0];
+                referenceTrajectoryIndices(T + 1) = Kt;
+                //Step 3:
+                //Compute residual ancestor indices via sampling from categorical distribution
+                if (l < N) {
+                    DsetDeterministic = arma::linspace<arma::Col<unsigned int> >(l + 1, N, N - l);
+                    arma::uvec DsetFinalIndices = arma::find(DsetDeterministic != Kt);
+                    DsetDeterministic = DsetDeterministic.elem(DsetFinalIndices);
+                    for(int k : DsetDeterministic) {
+                        uRSIndices.at(k) = Rcpp::sample(N, 1, false, Rcpp::wrap(dRSWeightsResidual))[0];
+                    }
+                }
+                //Step 4: If not done in 1. assign conditional ancestor index:
+                if(arma::all(DsetKtMinus1 != Kt)) {
+                    uRSIndices.at(Kt) = referenceTrajectoryIndices.at(T);
+                }
+                //else already assigned under 1.
+                break;
+            }
+        }
+        //Perform the replication of the chosen.
+        for(int i = 0; i < N ; ++i) {
+            if(uRSIndices(i) != static_cast<unsigned int>(i)){
+                pPopulation.SetValueN(pPopulation.GetValueN(static_cast<int>(uRSIndices(i))), i);
```

Comment:
I may be misunderstanding something here, but is there a possible race condition here? The value of each particle is set to the value of its ancestor in a loop but the value of the ancestor could already have been overwritten, no? (This was dealt with in the standard case by assuring that if particle i had any offspring then one of them had index i but I don't think that this is the case here.)

Am I missing something (I'm quite prepared to believe that I am)? Otherwise the cleanest option is to copy the pre-resampling particle set and use the values within that collection where we have `GetValueN` at present.

Reply:
Yes you are absolutely right! 

I've done a change though I am not sure if it is 100% what you suggested. Copying the whole particle might be too much (and also It think as currently implemented `=` does a shallow copy of the pPopulation object).

So I copied only `pPopulation.value` member via `GetValue()` into a seperate `populationValueCopy` and did the ancestor assignment with values taken from that vector.

We might want to add a note on the To-Do:
1. Either add a member that permutes the particle set to the `population` class (and then implement it efficiently later). The operation might be common enough to justify a member-function for that.
2. Permutation member function ccould be optimized. I __*think*__ there are some algorithmic ways how to permute without pre-copying that might be worth considering for large particle sets, but that is definitely something for the future too.

Reply:
Thanks. Indeed, I meant copy the collection of values rather than the object itself. 
I agree that 1, 2 might make sense -- and that we at least want the facility to perform deep copying, whether that's default behaviour (which perhaps it should be) or not.

Reply:
I think what you have here should be fine; but perhaps the copy should be explicitly cleared once we're finished with it?